### DISCUSSION TITLE
Is it worth to suggest to explicitly forbid declaration patterns for case of multiple `switch_label` in single `switch_section`

### DISCUSSION THREAD

**Discussion by gurustron** [dotnet/csharplang#7786] [Q&A] [answered]
As far as I can see currently specification allows any to be pattern used in `switch_label` no matter how many labels are present in single section (and compiler conforms):

> Each `switch_section` consists of one or more `switch_labels` followed by a `statement_list` (§13.3.2). Each `switch_label` containing case has an associated pattern ([§11](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/patterns#11-patterns-and-pattern-matching)) against which the value of the switch expression is tested.  

But for example using declaration pattern in case of multiple labels does not make much sense (as far as I understand) since the variable can't be used due to unassigned variable (for obvious reasons):

```
void Foo(object obj)
{       
	switch (obj)
	{
		case string s:
		case object o:
			Console.WriteLine(obj);
			// Console.WriteLine(s); // error CS0165: Use of unassigned local variable 's'
			// Console.WriteLine(o); // error CS0165: Use of unassigned local variable 'o'
			break;
	}      
}
```

P.S.

Seems to be pretty "edge-caseish" and probably not worth the effort.

**Comment by HaloFour** [comment, marked as answer]
I forget the name of the proposal (can look it up later) but there is one for allowing overlap of the pattern variables between disjunctive patterns that would allow something like this:

```cs
switch (obj) {
    case Person(string name):
    case Business(string name):
        // can use name here, it's definitely assigned by one of the two patterns
        break;
}
```

And while those variables aren't definitely assigned, they could be within the switch block and then they could be used.

**Reply by ufcpp** [reply]
https://github.com/dotnet/csharplang/issues/4018 ?
